src/auth.rs

```diff
@@ -87,3 +90,172 @@ impl AppAuth {
+    /// "ACCEPT: application/json". For example:
+    /// ```no_run
+    /// # async fn run() -> octocrab::Result<()> {
+    /// # use reqwest::header::ACCEPT;
+    /// let crab = octocrab::Octocrab::builder()
+    /// .base_url("https://github.com")?
+    /// .add_header(ACCEPT, "application/json".to_string())
+    /// .build()?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub async fn authenticate_as_device<I, S>(
+        &self,
+        client_id: &SecretString,
+        scope: I,
+    ) -> Result<DeviceCodes>
+    where
+        I: IntoIterator<Item = S>,
+        S: AsRef<str>,
+    {
+        let scope = {
+            let mut scopes = scope.into_iter();
+            let first = scopes
+                .next()
+                .map(|s| s.as_ref().to_string())
+                .unwrap_or_default();
+            scopes.fold(first, |i: String, n| i + "," + n.as_ref())
+        };
+        let codes: DeviceCodes = self
+            .post(
+                "login/device/code",
+                Some(&DeviceFlow {
+                    client_id: client_id.expose_secret(),
+                    scope: &scope,
+                }),
+            )
+            .await?;
+        Ok(codes)
+    }
+}
+
+/// The device codes as returned from step 1 of Github's device flow.
+///
+/// See https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#response-parameters
+#[derive(Deserialize, Clone)]
+#[non_exhaustive]
+pub struct DeviceCodes {
+    /// The device verification code is 40 characters and used to verify the device.
+    pub device_code: String,
+    /// The user verification code is displayed on the device so the user can enter the
+    /// code in a browser. This code is 8 characters with a hyphen in the middle.
+    pub user_code: String,
+    /// The verification URL where users need to enter the user_code: https://github.com/login/device.
+    pub verification_uri: String,
+    /// The number of seconds before the device_code and user_code expire. The default is
+    /// 900 seconds or 15 minutes.
+    pub expires_in: u64,
+    /// The minimum number of seconds that must pass before you can make a new access
+    /// token request (POST https://github.com/login/oauth/access_token) to complete the
+    /// device authorization. For example, if the interval is 5, then you cannot make a
+    /// new request until 5 seconds pass. If you make more than one request over 5
+    /// seconds, then you will hit the rate limit and receive a slow_down error.
+    pub interval: u64,
+}
+
+impl DeviceCodes {
+    /// Poll Github to see if authentication codes are available.
+    ///
+    /// See `https://docs.github.com/en/developers/apps/building-oauth-apps/authorizing-oauth-apps#response-parameters` for details.
+    pub async fn poll_once(
+        &self,
+        crab: &crate::Octocrab,
+        client_id: &SecretString,
+    ) -> Result<Either<OAuth, Continue>> {
```

Comment:
I personally like exposing `poll_once` only after `authenticate_as_device` has been called. The other option is exposing the function as 
```rust
Octocrab::poll_for_device_authentication(&self, codes: &DeviceCodes, client_id: &SecretString)
```

I'm not sure which one would be more idiomatic in the crate.